v2/log/sampling.go

```diff
@@ -37,6 +40,7 @@ func (s *sampleSyncer) Write(b []byte) (int, error) {
 	if s.count > s.first && s.count%s.thereafter != 0 {
 		return ioutil.Discard.Write(b)
 	}
+	s.Unlock()
```

Comment:
This solution have possibility to create a deadlock.
Run this at line 33 with defer instead - fixes the issue.

Reply:
Pushed a new commit that adds an unlock in the if-clause as well. Wanted to release it before writing to speed it up. Should I defer it instead anyway?

Reply:
I see, now you have fixed the issue, but still want it in a defer if any new developer want to add a new return in this function. Less chance of mistakes

Reply:
Or only lock it when changing or reading the count :D

Reply:
Good point! Deferring it now instead.

Reply:
Maybe you should break this out to a new function(s). :) We only want it to block the s.count